design/ceph/object/ceph-rgw-k8s-sa-authentication.md

```diff
@@ -0,0 +1,66 @@
+---
+Service authentication with vault for RGW
+target-version: release-1.10
+---
+
+# Feature Name
+Service account authentication with vault for RGW 
+
+## Summary
+When `Vault` is set up in Kubernetes environment, it can be authenticated using service account token. The OSD encryption already supports this feature and adding same for RGW. It uses sameset of  config option in `ConnectionDetails` for `KeyManagementServiceSpec`.
+
+### Goals
+Configure RGW with service account authentication with Vault with help of vault agent.
+
+### Non-Goals
+There are different ways RGW can authenticate with help of [vault agent](https://docs.ceph.com/en/latest/radosgw/vault/#vault-agent), but only service account authentication is supported here.
+
+## Proposal details
+The service account details will specified in `ConnectionDetails` of `KeyManagementServiceSpec`, using that info plus the other details a config map with <ceph-object-store>-vault-agent-cm populated for the vault agent sidecar container. The side container details will be added RGW pod spec and start sidecar container with RGW pod. The RGW will be configured with vault agent specific options. For the following details in `ConnectionDetails`:
+```
+security:
+  kms:
+    connectionDetails:
+        KMS_PROVIDER: vault
+        VAULT_ADDR: https://vault.default.svc.cluster.local:8200
+        VAULT_BACKEND_PATH: rook
+        VAULT_SECRET_ENGINE: kv
+        VAULT_AUTH_METHOD: kubernetes
+        VAULT_AUTH_KUBERNETES_ROLE: rook-ceph
+```
+The configmap for vault agent will be:
+```yaml
+kind: ConfigMap
+metadata:
+  name: <ceph-object-store>-vault-agent-cm
+  namespace: rook-ceph
+apiVersion: v1
+data:
+  vault-agent-config.hcl: |
+    pid_file = "/home/vault/pidfile"
+
+    auto_auth {
+        method "kubernetes" {
+            mount_path = "auth/kubernetes"
+            config = {
+                role = "rook-ceph"
+            }
+        }
+   }
+   cache {
+        use_auto_auth_token = true
+   }
+   listener "tcp" {
+        address = "127.0.0.1:8100"
+        tls_disable = "true"
+   }
+```
+### Risks and Mitigation
+User will be able to modify the `vault-agent-cm` which is not preferable
+
+## Config Commands
+The user can itself bring up `vault agent` as separate deployment configuring with service account authentication details. Then set following for RGW via toolbox pod:
```

Comment:
If the vault agent can be started as a separate deployment, is only one agent needed for the whole cluster? If so, should we anyway create a vault agent deployment instead of an rgw sidecar? Or why should it be a sidecar for rgw?

Reply:
Good question!

Reply:
I have done a fair bit more research on this. I believe that Vault's model that the vault agent runs as a sidecar is merely a requirement because it assumes the agent injects secrets into a shared pod volume. If RGW is talking to vault agent directly, there is no need to run the agent as a sidecar, because there is no need to have the agent inject anything into pod shared directories.

This means we can have (and capture in the design doc) the decision to either use a sidecar or to use a standalone vault agent that is shared by all RGWs. 

Using sidecars will naturally increase our resource footprint, a negative. However, the RGW can contact the sidecar on `localhost` within the Pod.

Using a standalone agent will decrease resources, especially for large numbers of RGWs. However, RGWs will likely be run on many nodes, and a shared agent will add additional host-to-host latency to each S3 call IIUC.

-----

Regarding the agent injector, while it is intended to be used to inject specific secrets into Pods, I believe it has the flexibility to configure the agent as a listener as well by modifying the ConfigMap as documented here: https://www.vaultproject.io/docs/platform/k8s/injector/examples#configmap-example

We can set the [listener config](https://www.vaultproject.io/docs/agent#listener) in the ConfigMap to have the agent sidecar automatically injected and listening on a port expected by the RGW.

In my opinion, a better design may be to allow users the flexibility to choose whether they want to run a standalone vault agent or whether they want sidecars. It would be easy to add an option to the CephObjectStore `security` spec that defines the address of the vault agent: `spec.security.vaultAgentHostname`.

If they want **sidecars**, then they can...
1. set `spec.security.vaultAgentHostname` to `localhost:<port>`
2. configure the agent `listener` to `localhost:<port>`
3. add the below annotations on the CephObjectStore.
    ```yaml
      annotations:
        vault.hashicorp.com/agent-inject: 'true'
        vault.hashicorp.com/agent-configmap: 'my-configmap' on the CephObjectStore and how to set the agent 
    ```
    
If they want to use a **standalone** vault agent, then they merely need to set `spec.security.vaultAgentHostname` to the Service that provides the vault agent.

Reply:
From the vault docs, `listener` specifies the address on which the vault agent listens. I believe this means we can (using the confimap) tell the agent that is injected to listen to RGW requests in the pod at something like `localhost:6543`. We can then configure `vaultAgentHostname` to be `localhost:6543`.

Reply:
@BlaineEXE: Yaa I can understand `listerner` with vault-agent, sorry I was confused about how it can be related to `vault-injector` The PR https://github.com/rook/rook/pull/9872https://github.com/rook/rook/pull/9872 have a similar configuration or approach